### PR TITLE
[Kwan Hao Wei] iP

### PR DESCRIPTION
**Pull request by KwanHW**
# Waldo

> Can't find me if I'm on Quarantine Order!

Waldo frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1.  download it from [here](https://github.com/KwanHW/ip/releases).
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x]     Managing tasks
- [x]     Managing deadlines
- [ ]     Reminders (coming soon)

<hr>
If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
   public static void main(String[] args) {
        new Duke();
        Application.launch(Duke.class, args);
    }
```

### CODE REVIEW

File: src/test/java/duke/task/TaskStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        assertTrue(new TaskStore().getIsEmpty(), "Failed: Task Store not empty when initialised");
+    }
+
+    @Test
+    public void noTasksOnDate() {
+        TaskStore ts = new TaskStore();
+        LocalDate date = LocalDate.parse("2022-01-31");
+        ts.addTask(new Deadline("Complete Assignment", date));
+        ts.addTask(new Event("Project Meeting", date));
+        assertEquals("You don't have any tasks on Jan 29 2022", ts.getTasksOn(LocalDate.parse("2022-01-29")), "Failed: TaskStore contains tasks that do not fall on that date");
```
Comment (LapisRaider) [inline src/test/java/duke/task/TaskStoreTest.java:22]: Line length longer than 120 characters, might want to warp it.

Comment (LapisRaider) [inline src/test/java/duke/task/TaskStoreTest.java]: Line length longer than 120 characters, might want to wrap it.

File: src/test/java/duke/task/TaskStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Test
+    public void noTasksOnDate_noTimeables() {
+        TaskStore ts = new TaskStore();
+        LocalDate date = LocalDate.parse("2022-01-31");
+        ts.addTask(new Todo("Complete Assignment"));
+        ts.addTask(new Todo("Project Meeting"));
+        assertEquals("You don't have any tasks on Jan 29 2022", ts.getTasksOn(LocalDate.parse("2022-01-29")), "Failed: TaskStore only contains non-timeable instances.");
```
Comment (LapisRaider) [inline src/test/java/duke/task/TaskStoreTest.java]: Line length longer than 120 characters, might want to wrap it.

File: src/test/java/duke/task/TaskStoreTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Test
+    public void tasksOnDate() {
+        TaskStore ts = new TaskStore();
+        LocalDate date = LocalDate.parse("2022-01-29");
+        ts.addTask(new Deadline("Complete Assignment", date));
+        ts.addTask(new Event("Project Meeting", date));
+        String expected = "Here are your tasks on Jan 29 2022\n" + ts.getTask(0).toString() + "\n" + ts.getTask(1).toString() + "\n";
```
Comment (LapisRaider) [inline src/test/java/duke/task/TaskStoreTest.java]: Line length longer than 120 characters, might want to wrap it. There might be other lines longer than 120 characters that I couldn't catch. 

IntelliJ has this formatting option that will help you wrap the lines longer than a certain amount of characters (you can set the limit in settings). They have a bunch of other settings too, definitely a time saver instead of manually scanning through every line of the codes!

File: src/main/java/duke/task/Task.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public String writeToFile() {
+        String isDone = this.isDone ? "1" : "0";
+        return String.join(" | ", this.taskType.toString(), isDone, this.description);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[%s][%s] %s", this.taskType.toString(), this.getStatusIcon(), this.description);
+    }
```
Comment (LapisRaider) [inline src/main/java/duke/task/Task.java]: The usage of `this.` keyword. It is recommended to use the `this` keyword only when a field is shadowed by a method or constructor parameter.

image

File: src/main/java/duke/task/Event.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents an event that is happening on a certain date.
+ */
+public class Event extends Task implements Timeable {
+    protected LocalDate at;
```
Comment (LapisRaider) [inline src/main/java/duke/task/Event.java]: Might want to change this to private instead of protected. For java, unlike C++ where protected makes the variable inaccessible to other classes besides the derived classes, in java, as long as a class is in the same package, they can access the protected variable.

If that is not your intention, you may want to change to private since Event class is only using this variable (correct me if I'm wrong).

File: src/main/java/duke/task/Deadline.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents a task that has to be completed before a certain date.
+ */
+public class Deadline extends Task implements Timeable {
+    protected LocalDate by;
```
Comment (LapisRaider) [inline src/main/java/duke/task/Deadline.java]: Might want to change this to private instead of protected. For java, unlike C++ where protected makes the variable inaccessible to other classes besides the derived classes, in java, as long as a class is in the same package, they can access the protected variable.

If that is not your intention, you may want to change to private since Deadline class is only using this variable (correct me if I'm wrong).

File: src/main/java/duke/io/Storage.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
+package duke.io;
+
+import duke.task.*;
```
Comment (LapisRaider) [inline src/main/java/duke/io/Storage.java]: Might want to list out the imports explicitly instead of using wildcard imports since it's not recommended.

**Review by LapisRaider — COMMENTED** [review]
Looks great! Just a few nits to fix.

TLDR:
- `this` keyword is recommended being used only when a field is shadowed by a method or constructor parameter.
- Some of your lines might be longer than 120 characters, IntelliJ can help you format.
- Protected might want to change to private. Protected keyword in Java makes the variable/function accessible for any class in the same package.
- Might want to change wildcard imports to listing explicitly instead.

File: src/main/java/duke/task/Event.java (hunk 5 above)
Comment (KwanHW) [reply src/main/java/duke/task/Event.java]: Thank you for catching that error. I may have gotten it mixed up with experience in previous languages (like C#). 

I have resolved the other issues similar as the one mentioned
